src/raw_node.rs

```diff
@@ -397,14 +439,11 @@ impl<T: Storage> RawNode<T> {
             self.commit_since_index = e.get_index();
         }
 
-        if !self.messages.is_empty() {
-            mem::swap(&mut rd.messages, &mut self.messages);
-        }
-
-        if raft.state == StateRole::Leader && !raft.msgs.is_empty() {
+        if !raft.msgs.is_empty() {
+            rd.messages = mem::take(&mut raft.msgs);
             // Leader can send messages immediately to make replication concurrently.
             // For more details, check raft thesis 10.2.1.
-            rd.messages.push(mem::take(&mut raft.msgs));
+            rd.is_persisted_msg = raft.state != StateRole::Leader;
```

Comment:
This implicitly requires followers to not step any messages before last `Ready` is advanced. This requirement should be documented in `ready` function.